src/server/checkkey.c

```diff
@@ -170,3 +174,409 @@ check_license(struct license_block *licenses)
+			pjob->ji_qs.ji_jobid, log_buffer);
+
+	} else {
+		licenses.lb_aval_floating += pjob->ji_licalloc;
+	}
+	update_FLic_attr();
+
+	sprintf(log_buffer,
+		"Released %d cpu licenses, float avail global %d, float avail local %d, used locally %d",
+		pjob->ji_licalloc, licenses.lb_glob_floating,
+		licenses.lb_aval_floating, licenses.lb_used_floating);
+	DBPRT(("%s\n", log_buffer))
+	log_event(PBSEVENT_DEBUG3, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+		pjob->ji_qs.ji_jobid, log_buffer);
+	pjob->ji_licalloc = 0;
+#endif
+}
+
+/**
+ * @brief
+ * 		clear_and_populate_svr_unlicensedjobs - removes all job entries from the
+ *		svr_unlicensedjobs list, and re-populates the list
+ *		with all currently running/exiting jobs (not suspended jobs).
+ *
+ */
+
+void
+clear_and_populate_svr_unlicensedjobs(void)
+{
+	job *pjob;
+
+	while ((pjob = GET_NEXT(svr_unlicensedjobs)))
+		delete_link(&pjob->ji_unlicjobs);
+
+	/* repopulate the list */
+	pjob = (job *)GET_NEXT(svr_alljobs);
+
+	while (pjob  != (struct job *)0) {
+
+		if (((pjob->ji_qs.ji_state == JOB_STATE_RUNNING) &&
+			((pjob->ji_qs.ji_svrflags & JOB_SVFLG_Suspend) == 0)) &&
+			!is_linked(&svr_unlicensedjobs, &pjob->ji_unlicjobs)) {
+			append_link(&svr_unlicensedjobs, &pjob->ji_unlicjobs,
+				pjob);
+			pjob->ji_licalloc = 0;
+		}
+		pjob = GET_NEXT(pjob->ji_alljobs);
+	}
+
+}
+
+
+/**
+ * @brief
+ * 		relicense_svr_unlicensedjobs - go through the svr_unlicensedjobs list
+ *   	and attempt to obtain as much licenses available to relicense the jobs.
+ * @note
+ * 		NOTE: Before calling this function, you may invoke the
+ *       clear_and_populate_svr_unlicensedjobs() function first,
+ *       so that the list is populated with jobs that are currently running.
+ *       The list is also updated by the calls to set_nodes() in node_manager.c.
+ *
+ * @par Algorithm:
+ *      For each job in svr_unlicensedjobs list,
+ *		calculate # of cpu licenses still needed by the job (job_needs)
+ *              If job_need <= 0  then remove job from the svr_unlicensedjobs
+ *                    list
+ *              Otherwise,
+ *                    See if there are usable licenses
+ *                     (# avail in local pool + # avail in global license pool)
+ *                   available. If so, then
+ *                   Attempt to satisfy the job by assigning it:
+ *                           MIN(job_needs, # of usable licenses)
+ *                   checking out as much additional licenses from the global
+ *                   license pool needed to satisfy the job.
+ *                   If satisfied, set job->licalloc to the actual # of
+ *                       assigned.
+ *                   If job->licalloc  becomes equal to the # of cpus
+ *                      requested by the job, then
+ *                             the job is fully licensed so it is
+ *                             removed from the svr_unlicensedjobs list.
+ *       done
+ */
+void
+relicense_svr_unlicensedjobs(void)
+{
+#if 0
```

Comment:
Why are we adding disabled code? Shouldn't the conditional be based on parameters passed to configure?